### PR TITLE
CMR-4479 Add support for concept-id service lookup

### PR DESCRIPTION
**Pull request by oubiwann**
Summary:

This adds the service concept as a valid type when querying the `/search/concepts` CMR resource.

Previous comment, during the WIP PR for this branch:

@ygliuvt I'm getting a null exception when running the new tests against the `/search/concepts` resource with a service concept it. I haven't been able to track down where the error is coming from. I'm wondering if something is missing from our service work for this to be supported? (i.e., another ticket) or if there's some part of the CMR that needs to be updated that produces this sort of side effect (e.g., an ACL update somewhere ... I seem to remember we did something like that with UMM-Vars ...).

### CODE REVIEW

File: search-app/src/cmr/search/services/acl_service.clj [hunk 1]
```diff
@@ -23,15 +23,36 @@
...
+;; When plans solidify around acl checks for service concepts, we'll update this.
+(defmethod acls-match-concept? :service
+  [context acls concept]
+  true)
+
 (defmethod acls-match-concept? :default
   [context acls concept]
   false)
 
+;; XXX This code is a point of contention for adding new concepts: on both
```
Comment (oubiwann) [inline search-app/src/cmr/search/services/acl_service.clj]: This is the second time this has bitten us. We need to incorporate this into our future plans for refactoring the code for adding new concepts to the CMR.

File: system-int-test/test/cmr/system_int_test/search/service/concept_retrieval_test.clj [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+          (testing "retrieval by service concept-id returns the latest revision."
+            (is (= svc1-name-new
+                   (:Name (json/parse-string (:body response) true)))))
+          (testing (str "retrieval by service concept-id and revision-id returns "
+                    "the specified service")
+            (is (= svc1-name
+                   (:Name (json/parse-string (:body response-v1) true))))
+            (is (= svc1-name-new
+                   (:Name (json/parse-string (:body response-v2) true)))))))
+    (testing "retrieval by service concept-id and incorrect revision-id returns error"
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/search/service/concept_retrieval_test.clj]: Combine this test case and the next one into one with `are3`.

File: system-int-test/test/cmr/system_int_test/search/service/concept_retrieval_test.clj [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            response-v1 (search/retrieve-concept
+                         (:concept-id svc1-v1)
+                         1
+                         {:accept mt/any})
+            response-v2 (search/retrieve-concept
+                         (:concept-id svc1-v1)
+                         2
+                         {:accept mt/any})]
+          (testing "retrieval by service concept-id returns the latest revision."
+            (is (= svc1-name-new
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/search/service/concept_retrieval_test.clj]: How about creating a helper function (e.g. `verify-retrieved-concept`) to do the whole retrieving concept, parsing JSON and comparing service name in just one function?

File: search-app/src/cmr/search/services/acl_service.clj [hunk 4]
```diff
@@ -23,15 +23,36 @@
   [context acls concept]
   true)
 
+;; When plans solidify around acl checks for service concepts, we'll update this.
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/acl_service.clj]: Needs a ticket number. If we don't have a ticket number for ACLs for services we need to write one.

File: search-app/src/cmr/search/services/acl_service.clj [hunk 5]
```diff
@@ -23,15 +23,36 @@
...
+;;     `acls-match-concept?` multimethod.
+;;
+;;     This needs to be taken into account when we plan for refactoring the
+;;     APIs and development process for adding concepts to the CMR. The fact
+;;     that this crucial change is tucked away in this part of the code with no
+;;     logical "pointers" or hints to it anywhere else is deeply problematic ...
+;;     programming by conventions is a slow path to developer insanity and
+;;     software project failure! Instead, all parts of the code that need to be
+;;     touched should be replaced with APIs (e.g., protocols and their impl'ns)
+;;     with each method documented and, all toegher, being the only things a
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/acl_service.clj]: Define _toegher_: Adjective, tougher and all together.

File: system-int-test/src/cmr/system_int_test/utils/service_util.clj [hunk 6]
```diff
@@ -18,6 +18,10 @@
 (def default-opts {:accept-format :json
                    :content-type content-type})
 
+(defn token-opts
+  [token]
```
Comment (chris-durbin) [inline system-int-test/src/cmr/system_int_test/utils/service_util.clj]: Add docstring

File: system-int-test/test/cmr/system_int_test/search/service/concept_retrieval_test.clj [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  [(ingest/reset-fixture {"provguid1" "PROV1"})
+   service/grant-all-service-fixture]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;
+;; Helper utility functions
+;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn- get-svc
```
[the remark this answers is not in the document]
Comment (oubiwann) [reply system-int-test/test/cmr/system_int_test/search/service/concept_retrieval_test.clj]: Both this one and the next one were created (copied here; created in the variable concept retrieval test) because the `let` blocks that contained them were getting unruly: long, nested calls inside them, side-effect operations that returned no useful value (causing us to use the `[_ (call)]` anti-pattern), etc.

But I simultaneously harbour objections similar in nature to yours: does this mean that we then have little one-off functions scattered around our testing codebase whose sole purpose it to help us keep clean code? This would indicate that a symptom was addressed, not a root problem.

Which begs the question: what are we not taking into consideration with our APIs such that its usage is so awkward? A good API has very little of these usage issues; an excellent API has none of them.

In the mean time, I'll take another look and see if there's a nicer way to clean these up here...

File: search-app/src/cmr/search/services/acl_service.clj (hunk 4 above)
Comment (oubiwann) [reply search-app/src/cmr/search/services/acl_service.clj]: Yeah, Liu and I looked at this comment this morning (it's a copy-paste) and where were wondering where it came from and what it might be referring to. I think I may have actually written the original in the variables version of this sys-int test ... I'll dig and try to remember, and then update the comment.

File: search-app/src/cmr/search/services/acl_service.clj (hunk 5 above)
Comment (oubiwann) [reply search-app/src/cmr/search/services/acl_service.clj]: I'll also reference the epic ticket here ...

Comment (oubiwann) [reply search-app/src/cmr/search/services/acl_service.clj]: Done.

File: system-int-test/test/cmr/system_int_test/search/service/concept_retrieval_test.clj (hunk 7 above)
[the remark this answers is not in the document]
Comment (ygliuvt) [reply system-int-test/test/cmr/system_int_test/search/service/concept_retrieval_test.clj]: I find these helper functions make the test hard to follow. I don't want my `get-svc` call to actually ingest a service concept with the given service-name and returns the ingested service concept id and revision id. I would rather see something like `ingest-service-with-attrs` which at least tells me a service concept is ingested and that info is more important to understand the test.

File: system-int-test/src/cmr/system_int_test/utils/service_util.clj (hunk 6 above)
Comment (oubiwann) [reply system-int-test/src/cmr/system_int_test/utils/service_util.clj]: Added.

File: search-app/src/cmr/search/services/acl_service.clj (hunk 4 above)
Comment (oubiwann) [reply search-app/src/cmr/search/services/acl_service.clj]: I'm having vague memories of the comment origins; I think I made them after a pairing session with Liu (the first time we ran into the testing issues as a result of this code not being updated when variable concepts were added). I think that comment came from me poorly paraphrasing something Liu said on the call, so I've updated it to essentially say the same thing that the tag comment says: basically, there's no ACLs for variables and services right now.

In other words, done ;-)

File: system-int-test/test/cmr/system_int_test/search/service/concept_retrieval_test.clj (hunk 7 above)
[the remark this answers is not in the document]
Comment (oubiwann) [reply system-int-test/test/cmr/system_int_test/search/service/concept_retrieval_test.clj]: Yeah, that's a good point; those helper function names hide a benefit of the long names: self-documentation.  

We might be able to address my other concerns in the sys-int utility namespaces for now (and then, ultimately, a well-designed client library ... or utility library built on the client library).
